### PR TITLE
*: now lease keepAlive works on leader

### PR DESCRIPTION
**Pull request by xiang90**
Now the lease API is finished. 

We still need to make follower forward lease renew request to leader. Also we need to add more integration testing around lease API, especially when there is a leader failover.

### CODE REVIEW

**Comment by gyuho** [comment]
lgtm, codewise :)

File: etcdserver/api/v3rpc/lease.go [hunk 1]
```diff
@@ -41,5 +44,24 @@ func (ls *LeaseServer) LeaseRevoke(ctx context.Context, rr *pb.LeaseRevokeReques
...
+
+		ttl, err := ls.le.LeaseRenew(lease.LeaseID(req.ID))
+		if err != nil {
+			return ErrLeaseNotFound
+		}
+
+		resp := &pb.LeaseKeepAliveResponse{ID: req.ID, TTL: ttl}
+		err = stream.Send(resp)
+		if err != nil {
+			return nil
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/lease.go]: why nil instead of err?

File: etcdserver/api/v3rpc/lease.go [hunk 2]
```diff
@@ -41,5 +44,24 @@ func (ls *LeaseServer) LeaseRevoke(ctx context.Context, rr *pb.LeaseRevokeReques
...
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		ttl, err := ls.le.LeaseRenew(lease.LeaseID(req.ID))
+		if err != nil {
+			return ErrLeaseNotFound
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/lease.go]: this will mask ErrNotPrimary

File: etcdserver/v3demo_server.go [hunk 3]
```diff
@@ -35,8 +35,12 @@ type RaftKV interface {
 }
 
 type Lessor interface {
+	// create, revoke needs to commit through raft
 	LeaseCreate(ctx context.Context, r *pb.LeaseCreateRequest) (*pb.LeaseCreateResponse, error)
 	LeaseRevoke(ctx context.Context, r *pb.LeaseRevokeRequest) (*pb.LeaseRevokeResponse, error)
+
+	// renew is a local operation
```
Comment (heyitsanthony) [inline etcdserver/v3demo_server.go]: note this function returns a ttl

File: etcdserver/api/v3rpc/lease.go (hunk 1 above)
Comment (xiang90) [inline etcdserver/api/v3rpc/lease.go]: my mistake. will fix. sorry.

File: etcdserver/api/v3rpc/lease.go (hunk 2 above)
Comment (xiang90) [inline etcdserver/api/v3rpc/lease.go]: ErrNotPrimary is a temp error. We should remove it after we implement forwarding logic. But I agree that we should never mask it. Probably we can panic it for now.

File: etcdserver/v3demo_server.go (hunk 3 above)
Comment (xiang90) [inline etcdserver/v3demo_server.go]: sure.

**Comment by xiang90** [comment]
@heyitsanthony Fixed. PTAL.

**Comment by heyitsanthony** [comment]
lgtm

**Pull request merged**